specs/waku/v2/waku-v2-rpc-api.md

```diff
@@ -0,0 +1,459 @@
+| `method`  | Contains the JSON-RPC method that is being called   |
+| `params`  | An array of parameters for the request              |
+| `id`      | The request ID                                      |
+
+## Types
+
+In this specification, the primitive types `Boolean`, `String`, `Number` and `Null`, as well as the structured types `Array` and `Object`, are to be interpreted according to the [JSON-RPC specification](https://www.jsonrpc.org/specification#conventions). It also adopts the same capitalisation conventions.
+
+The following structured types are defined for use throughout the document:
+
+### WakuMessage
+
+Refer to [`Waku Message` specification](https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-message.md) for more information.
+
+`WakuMessage` is an `Object` containing the following fields:
+| Field | Type | Inclusion | Description |
+| ----: | :--: | :--: |----------- |
+| `payload` | `String` | mandatory | The message payload |
+| `contentTopic` | `Number` | optional | Message content topic for optional content-based filtering |
+| `version` | `Number` | optional | Message version. Used to indicate type of payload encryption. Default version is 0 (no payload encryption). |
+
+## Method naming
+
+The JSON-RPC methods in this document are designed to be mappable to HTTP REST endpoints. Method names follow the pattern `<method_type>_waku_<protocol_version>_<api>_<api_version>_<resource>`
+
+- `<method_type>`: prefix of the HTTP method type that most closely matches the JSON-RPC function. Supported `method_type` values are `get`, `post`, `put`, `delete` or `patch`.
+- `<protocol_version>`: Waku version. Currently **v2**.
+- `<api>`: one of the listed APIs below, e.g. `store`, `debug`, or `relay`.
+- `<api_version>`: API definition version. Currently **v1** for all APIs.
+- `<resource>`: the resource or resource path being addressed
+
+The method `post_waku_v2_relay_v1_message`, for example, would map to the HTTP REST endpoint `POST /waku/v2/relay/v1/message`.
+
+## Debug API
+
+### Types
+
+The following structured types are defined for use on the Debug API:
+
+#### WakuInfo
+
+`WakuInfo` is an `Object` containing the following fields:
+| Field | Type | Inclusion | Description |
+| ----: | :--: | :--: |----------- |
+| `listenStr` | `String` | mandatory | Address that the node is listening for |
+
+### `get_waku_v2_debug_v1_info`
+
+The `get_waku_v2_debug_v1_info` method retrieves information about a Waku v2 node
+
+#### Parameters
+
+none
+
+#### Response
+- [**`WakuInfo`**](#WakuInfo) - information about a Waku v2 node
+
+## Relay API
+
+Refer to the [Waku Relay specification](https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-relay.md) for more information on the relaying of messages.
+
+### Types
+
+The following structured types are defined for use on the Relay API:
+
+#### WakuRelayMessage
+
+`WakuRelayMessage` is an `Object` containing the following fields:
```

Comment:
What's the rationale for keeping this separate from WakuMessage? Similar to the keeping distinct types rationale in Nim/JSON RPC?

Reply:
Might want to explicitly write "this corresponds to a WakuMessage" or similar

Reply:
Similar rationale, but with the added trickery that, on the API, you may not necessarily want the messages you `post` to be the same as those you `get`:
- to `post` a RelayMessage, you don't (necessarily) want the client to set the `WakuMessage` `version` and handle corresponding encryption itself. There should rather be separate endpoints specified with/without encryption and the API logic should convert that to a corresponding `WakuMessage` with correct `version`, etc.
- to `get` messages, say from the WakuStore, a client may want to receive the full `WakuMessage` in order to interpret the payloads, especially if there are messages of different versions on the same content topic. A separate endpoint to just retrieve decrypted messages will be useful here too, but it will still imply defining a separate type that suppresses `version` for clarity's sake.